Feature/throw exception when response status code not 200

4c0n opened the pull request:
Currently if the httplug adapter is not configured to throw exceptions on non 200-ok statuses, there is no way to check if a request to the slack api was successfull. In my case the calls were silently failing, providing no means of finding out what went wrong.

Now when http_errors is configured on the adapter, the behavior is the same as before, but when it is not configured to throw exceptions, this library will throw an exception.

4c0n committed:
Throwing package specific exception when the status code is not 200

nedac-sorbo committed:
added missing semi colon

soullivaneuh commented:
Hi @4c0n and thank you for your contribution. :+1: 

I'm not sure handling HTTP errors goes to the slack library I'm afraid, this is also why httplug provide a plugin to manage it on side project.

Could you please give us some use case with the errors you encounter?

Thanks

4c0n commented:
hi @soullivaneuh , thanks for your response!

For example I was using the wrong channel name, using the guzzle 6 adapter, with `http_errors: false`. The API responded with a 404 response, but there was no way to find out from this package. I had to debug using httplug to find out what was going wrong. Preferably the method either always returns `false` on failure or it should throw an exception.
Since using httplug is basically like a black box to the user (we're using this library not connecting to the api directly using httplug), I feel there should be a way to determine if sending the message was successful or not, without having to configure httplug to throw exceptions on none 200 statuses (but if that is what you really want, then it should be part of the documentation/instructions on how to use the library IMO).

soullivaneuh commented on src/SlackApiException.php:
```diff
@@ -0,0 +1,21 @@
+
+namespace Nexy\Slack;
+
+use Exception;
```
Please directly use `\Exception`

soullivaneuh commented on src/Client.php:
```diff
@@ -144,7 +144,11 @@ public function sendMessage(Message $message): void
-        $this->httpClient->post('', [], $encoded);
+        $response = $this->httpClient->post('', [], $encoded);
+        $statusCode = $response->getStatusCode();
+        if ($statusCode !== 200) {
```
- Yoda condition required.
- Also, a blank line before the `if` would be better.

soullivaneuh reviewed (changes requested):
I'm not sure. API related error should be indeed managed by this library, but not the network one. Especially if we move later to the new [HTTP Client PSR](https://www.php-fig.org/psr/psr-18/) and not directly httplug.

In that case, the assertion should be more accurate than "The code is different than 200". Maybe Slack has a documented error list about that?

Also, exception may be diversified, but having one is already a good start.

Do you agree?

I also did some code review. Could you please add an input on the changelog? https://github.com/nexylan/slack/blob/master/CHANGELOG.md

soullivaneuh commented:
Note: If merged, it must be released as a new major. Throwing exception is a BC break.

Commit:
Using Exception without importing it first, as requested in review

Commit:
Added more diverse exceptions; for every api error documented at https://api.slack.com/changelog/2016-05-17-changes-to-errors-for-incoming-webhooks

Commit:
Formatted following project standards

Commit:
Added phpunit to dev dependencies

Commit:
Added ErrorResponseHandler to handle slack api errors

Commit:
Removed final keyword so that the class is mockable in the tests

Commit:
Using ErrorResponseHandler in Client

Commit:
Added entry to changelog as requested

4c0n commented:
@soullivaneuh Thanks for the review! I looked up the API errors and implemented exceptions for each of them and added an `ErrorResponseHandler` class.
Please have another look and let me know if there is anything you would like to do differently. I will be testing it shortly.

Commit:
Using stream

Commit:
Using response body text as exception message

Commit:
Passing status code as exception code

4c0n commented:
@soullivaneuh I tested the changes and made some adjustments. When httplug is not configured to throw exceptions, this library will now throw exceptions when an API error occurs.
Potential changes/improvements:
- Throw IOException from ErrorResponseHandler for errors unrelated to the API
- Catch exceptions from httplug and process them as well to see if there was an API error if httplug IS configured to throw exceptions

By implementing the above we can guarantee that the behavior is the same regardless of the exception throwing configuration of httplug. The original exception could be included in the library one. So the library would not be handling http errors, just passing them on, which IMO makes sense, as devs should be dealing with errors coming from this lib, not dependencies of this lib.

Please let me know how you feel about this so that we can move forward :tada: 

Merci beaucoup!

4c0n commented:
@soullivaneuh Friendly ping

soullivaneuh commented on composer.json:
```diff
@@ -1,4 +1,4 @@
-{
+    {
```
Not wanted indent.

soullivaneuh commented on composer.json:
```diff
@@ -26,7 +26,8 @@
         "php-http/message": "^1.6",
-        "php-http/mock-client": "^1.2"
+        "php-http/mock-client": "^1.2",
+        "phpunit/phpunit": "^8.2"
```
Please do not require phpunit as a project dependency.

4c0n replied:
How can we run the unit tests, without PHPUnit? It is a dev dependency so it does not get in the way. Also there is a dependency on Mockery as well, is it different somehow?

Just curious, of course if you want I'll remove it, but I'd like to understand the reasoning.

soullivaneuh replied:
You have different way to install PHPUnit: https://phpunit.de/getting-started/phpunit-8.html

If you want IDE completion, I would recommend to use the .phar.

Reasons to not require it:

1. It's a dependency hell, making the project setup heavier.
2. When you install a .phar, you are sure it's a well tested and packaged version with exact dependency. With composer, PHPUnit might crash because of an update of one of its depencies
3. On some project, you might have some dependencies conflict with packages you want to require and phpunit.

The alternative way would be to use the [composer-bin](https://packagist.org/packages/bamarni/composer-bin-plugin) plugin, but it will not solve reason `2`.

OFC, the reasons are not shared by every one and I'm not pretending to bring the real truth here. :wink:

4c0n replied:
Got it, thanks for explaining :D

soullivaneuh commented on src/Client.php:
```diff
@@ -144,7 +159,11 @@ public function sendMessage(Message $message): void
+
+        $response = $this->httpClient->post('', [], $encoded);
+
+        // Throw exception if there is an API error, do nothing otherwise
```
This description should be and the handler class directly IMO.

soullivaneuh commented on src/ErrorResponseHandler.php:
```diff
@@ -0,0 +1,73 @@
+
+class ErrorResponseHandler
+{
+    // Mapping based on: "https://api.slack.com/changelog/2016-05-17-changes-to-errors-for-incoming-webhooks"
```
`@see https://...`

soullivaneuh commented on tests/ClientUnitTest.php:
```diff
@@ -17,7 +17,12 @@ class ClientUnitTest extends PHPUnit\Framework\TestCase
     {
-        $client = new Client('http://fake.endpoint', [], new \Http\Mock\Client());
+        $client = new Client(
+            Mockery::mock('Nexy\Slack\ErrorResponseHandler'),
```
Use `ErrorResponseHandler::class`

soullivaneuh commented on tests/ClientUnitTest.php:
```diff
@@ -36,7 +41,12 @@ public function testInstantiationWithDefaults(): void
 
-        $client = new Client('http://fake.endpoint', $defaults, new \Http\Mock\Client());
+        $client = new Client(
+            Mockery::mock('Nexy\Slack\ErrorResponseHandler'),
```
Use `::class` notation.

soullivaneuh commented on tests/ClientUnitTest.php:
```diff
@@ -55,7 +65,12 @@ public function testCreateMessage(): void
 
-        $client = new Client('http://fake.endpoint', $defaults, new \Http\Mock\Client());
+        $client = new Client(
+            Mockery::mock('Nexy\Slack\ErrorResponseHandler'),
```
Use `::class` notation.

soullivaneuh commented on tests/ErrorResponseHandlerUnitTest.php:
```diff
@@ -0,0 +1,147 @@
+{
+    private function getResponseMock(int $code, string $phrase, string $body): \Psr\Http\Message\ResponseInterface
+    {
+        $responseMock = Mockery::mock(\Psr\Http\Message\ResponseInterface::class);
```
Please import the interface.

soullivaneuh commented on src/ErrorResponseHandler.php:
```diff
@@ -0,0 +1,73 @@
+use Nexy\Slack\Exception\RollupErrorException;
+use Nexy\Slack\Exception\SlackApiException;
+use Nexy\Slack\Exception\UserNotFoundException;
+
```
Useless blank line.

soullivaneuh commented on tests/ClientUnitTest.php:
```diff
@@ -66,7 +81,12 @@ public function testCreateMessage(): void
     {
-        $client = new Client('http://fake.endpoint', [], new \Http\Mock\Client());
+        $client = new Client(
+            Mockery::mock('Nexy\Slack\ErrorResponseHandler'),
```
`::class` notation

soullivaneuh commented on CHANGELOG.md:
```diff
@@ -5,6 +5,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
 
 ## [Unreleased]
+### Changed
+- The `sendMessage()` method of the `Client` class will now throw exceptions that match
```
```markdown
- The `Client::sendMessage` method...
```

soullivaneuh reviewed (changes requested):
Sorry for the late reply.

Your approach looks great, thanks for the work!

I did some technical review.

RafaelFerreiraTVD commented:
I would like to see this on next release.
I have the same issue as @4c0n. Good job 😃

Commit:
Removed indentation

Commit:
Moved comment to doc block of relevant method

Commit:
Transformed comment into doc block

Commit:
Refactored to use ::class instead of namespace path as string

Commit:
Removed phpunit dependency

Commit:
Importing interfaces

Commit:
Using better changelog entry text

Commit:
Removed blank line

4c0n commented:
@RafaelFerreiraTVD Thanks!
@soullivaneuh I updated the PR following your requests, please have another look. :)

soullivaneuh commented on src/ErrorResponseHandler.php:
```diff
@@ -0,0 +1,77 @@
+class ErrorResponseHandler
+{
+    /**
+     * Mapping based on:
```
To be remove, the `@see` tag is enough. :+1:

soullivaneuh commented on src/Client.php:
```diff
@@ -41,12 +47,19 @@ final class Client
-    public function __construct(string $endpoint, array $options = [], HttpClient $httpClient = null)
-    {
+    public function __construct(
+        ErrorResponseHandler $errorResponseHandler,
```
Finally, I'm not sure it worth it to set it on the constructor. It's a single class and should not be replaced by something else.

Having a separated class for this is good to not make the client to heavy, but it should be instantiated directly into the constructor and not as a required parameter.

What do you think?

4c0n replied:
I feel it gives a better separation of concerns, the client is already over 200 lines of code now, so offloading some functionality to its own class is providing clarity and maintainability, following principles like SOLID. Also this allows for testing response handling in isolation of the rest of the process.

As far as it being heavy or not, Symfony will instantiate both classes pretty much at the same time and inject the ErrorResponseHandler into the client (unless defined as a lazy loaded service), I'm guessing that will give a similar overhead as to instantiating it in the constructor.
In any case the difference is probably not significant enough to state that one way is better than the other. 

However instantiating it in the constructor would take away the ability to test the classes in isolation. I think what you're looking for is something like an inner or friend class, but PHP doesn't have such class scopes.

For standalone usage of this package it would convenient not to have to pass the error handler to the client on each instantiation, we could add a Factory for that?

soullivaneuh replied:
> I feel it gives a better separation of concerns, the client is already over 200 lines of code now, so offloading some functionality to its own class is providing clarity and maintainability, following principles like SOLID

We agree about that, no problem. :+1: 

> However instantiating it in the constructor would take away the ability to test the classes in isolation.

I'm not sure to understand here. When you test the `Client`, you always have to passe a handler instance to it, so what is the difference instanciating it on the constructor?

4c0n replied:
The difference is that you are not able to pass a mock object instead of the real thing that way. Thus to reach 100% code coverage you would have to test the logic of the error handler together with the client, the tests will become more complex and perhaps even more fragile because of that.

Also I can imagine that at the moment this library is used for just the webhook API, but in the future you might want to support the other APIs as well, which will produce different errors, requiring a different error handler.

Anyway it is up to you, but we'd have to refactor the ClientUnitTest and instead of mocking the error handler we'd have to mock the response objects to make sure that the error handler does the right thing for each test.

4c0n replied:
We could use this, to overcome the isolation problem btw:
http://docs.mockery.io/en/latest/reference/creating_test_doubles.html#overloading

But it gets a bit tricky (fragile) when you need to retain the global state and need the error handler in a different test as well..

soullivaneuh replied:
> The difference is that you are not able to pass a mock object instead of the real thing that way.

This is true, but according to the concerning object, is it really worth it to mock it? If not, I don't think it worth it to force it as a constructor argument.

Commit:
Removed comment text, leaving only the @see annotation

Commit:
Instantiating error handler class in constructor of client

4c0n commented:
@soullivaneuh I modified the code to instantiate the error handler in the client constructor. Please have a look.

soullivaneuh commented on src/Client.php:
```diff
@@ -41,12 +46,17 @@ final class Client
-     * @param HttpClient|null $httpClient
+     * @param string               $endpoint
+     * @param array                $options
+     * @param HttpClient|null      $httpClient
```
I think you can remove the indentation now.

soullivaneuh reviewed (changes requested):
Looks good to me, just fix the CI and my last comment and we are good to merge!

Commit:
Fixed indentation

4c0n commented:
@soullivaneuh The php7.1 environment in Travis is running a no longer supported version of PHPUnit (6.x):
https://phpunit.de/supported-versions.html (did not check the other environments yet).

I never used Travis before and can't find a way to upgrade PHPUnit in the build system itself. Do you know if that is possible? I did read that it "prefers" phpunit versions that are installed through composer over the globally installed version. https://docs.travis-ci.com/user/languages/php/#default-build-script

Please advise.

soullivaneuh commented:
If you need a more recent version of PHPUnit, the simpler way would be to require it globally with composer, as the bin path is already setup on this project. If you just need the last stable version, then `composer global require phpunit/phpunit:@stable`.

The other way would be to simply download the `.phar` file: https://phpunit.de/getting-started/phpunit-8.html

Commit:
Globally installing the latest stable version of phpunit

4c0n committed:
Merge branch 'master' into feature/throw-exception-when-response-status-code-not-200

Commit:
Fixed code style

Commit:
Fixed code style

Commit:
Merge branch 'feature/throw-exception-when-response-status-code-not-200' of 4c0n.github.com:4c0n/slack into feature/throw-exception-when-response-status-code-not-200

4c0n commented:
@soullivaneuh The builds are fixed. Please have a last look.

4c0n commented:
@soullivaneuh Could you please have a look at this? Please keep in mind that this PR has been open for almost 3 weeks now. I would very much like to merge and close this PR as soon as possible.

soullivaneuh commented:
Hi @4c0n,

I understand, but please keep in mind it's the open-source world and I also have work to do for my company.

I will look at this at soon as possible.

Regards.

soullivaneuh commented:
Also, waiting any release, please note you can use your fork on your project dependencies! :+1:

soullivaneuh commented on .travis.yml:
```diff
@@ -39,6 +39,7 @@ install:
   - wget http://cs.sensiolabs.org/download/php-cs-fixer-v2.phar --output-document php-cs-fixer.phar
   - composer global update --prefer-dist --no-interaction
   - composer update --prefer-dist --no-interaction $COMPOSER_FLAGS
+  - composer global require phpunit/phpunit:@stable
```
Should be added on the `before_install` line, we already have a global require there.

soullivaneuh commented on CHANGELOG.md:
```diff
@@ -5,6 +5,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
 ## [Unreleased]
+### Changed
+- The `Client::sendMessage()` method will now throw exceptions that match
+API errors
```
One line only.

soullivaneuh reviewed (changes requested):
Two more requests, and we are good to go! :+1:

Commit:
Moved global phpunit installation to before_install step

Commit:
Using 1 line for changelog entry

soullivaneuh commented on .travis.yml:
```diff
@@ -34,6 +34,7 @@ cache:
 
 before_install:
   - composer global require satooshi/php-coveralls:@stable sllh/composer-lint:@stable --no-update
+  - composer global require phpunit/phpunit:@stable
```
I do not understand, why not just add `phpunit/phpunit:@stable` on the already present `global require` line?

RafaelFerreiraTVD replied:
Yeah, i agree @soullivaneuh
Just adding on last position of that line.

Commit:
Installing phpunit on the same line as other dependencies

RafaelFerreiraTVD commented:
Hi @soullivaneuh it seems good to merge.

soullivaneuh committed:
Add missing blank line

soullivaneuh commented:
Thank you for you work and you patience @4c0n! :+1:
